lexibank_lieberherrkhobwa.py

```diff
@@ -1,102 +1,121 @@
         value = self.lexemes.get(value, value)
-        return [self.clean_form(item, form).strip()
-                for form in split_text(value, separators='~/,;')]
 
-    def cmd_install(self, **kw):
+        # Return a list of cleaned forms, splitting over tilde, slash,
+        # comma, and colon
+        forms = [
+            self.clean_form(item, form).strip()
+            for form in split_text(value, separators='~/,;⪤')
+        ]
 
+        return forms
+
+
+    def cmd_install(self, **kw):
+        # Read the raw data
         data = self.raw.read_csv("dataset_khobwa.csv")
-        assert set(len(r) for r in data) == {201}
-        concept_by_index = OrderedDict()
-        for i in range(1, len(data[0]), 2):
-            concept_by_index[i] = data[0][i]
 
         with self.cldf as ds:
+            # Add bibliographic sources
             ds.add_sources(*self.raw.read_bib())
 
-            # Read raw concept data and append it
-            for concept in self.conceptlist.concepts.values():
-                for i, gloss in list(concept_by_index.items()):
-                    if gloss == concept.english:
-                        concept_by_index[i] = concept.id
-                        break
-                else:
-                    raise ValueError(concept["ENGLISH"])
+            # Read raw concept data and add to dataset; at the same time,
+            # build a map between the concept index in data and the
+            # concept id in the dataset
+            concept_by_index = {}
+            for cid, concept in enumerate(self.conceptlist.concepts.values()):
+                # We can get the concept id from the raw data
+                concept_by_index[1+(cid*2)] = concept.id
 
+                # Add the concept
                 ds.add_concept(
                     ID=concept.id,
                     Name=concept.english,
                     Concepticon_ID=concept.concepticon_id,
                     Concepticon_Gloss=concept.concepticon_gloss,
                 )
 
-            # Read raw languages and append it
-            langs = {}
-            for language in self.languages:
-                langs[language["Name"]] = language["Source"]
-                ds.add_language(
-                    ID=language["Name"],
-                    Name=language["Name"],
-                    Glottocode=language["Glottocode"],
-                    Glottolog_Name=language["Glottolog_Name"],
-                )
+            # Read raw languages and add to the dataset; at the same time,
+            # build a map that can be used for iterating over language ids
+            langs = {l['Name']:l['Source'] for l in self.languages}
+            ds.add_languages()
 
             # iterate over the source adding lexemes and collecting cognates
             for row in data[2:]:
+                # Get the language_id, which also allows to skip over
+                # data we don't want
                 lid = slug(row[0])
```

Comment:
You can slug() and not lowercase something. There's a keyword. You can check the call signature of the method (which would allow to use other IDs.

Reply:
It's `lowercase`, but I ended up adding a new `Source_Name` column in `etc/languages.csv`: the language identifier in the raw source is not necessarily a language name (like `zyymc` and `lidaqin`), and fixing with ifs would be ugly. Plus, having the name source in there seemed positive in terms of explicitness.